File: apps/web/src/features/sessions/procedures/proto/session.ts
import { timestampDate } from "@bufbuild/protobuf/wkt";
import { ORPCError } from "@orpc/client";
import {
	type Session,
	SessionsSchema,
} from "@/features/sessions/types";
import type { Session as SessionProto } from "@/gen/proto/planner/v1/session_pb";

export function protoToSession(proto: SessionProto): Session {
	if (!proto.createdAt) {
		throw new ORPCError("INTERNAL_SERVER_ERROR", {
			message: "Session is missing createdAt",
		});
	}
	if (!proto.updatedAt) {
		throw new ORPCError("INTERNAL_SERVER_ERROR", {
			message: "Session is missing updatedAt",
		});
	}
	if (!proto.startsAt) {
		throw new ORPCError("INTERNAL_SERVER_ERROR", {
			message: "Session is missing startsAt",
		});
	}

	return SessionsSchema.parse({
		campaignId: proto.campaignId,
		createdAt: timestampDate(proto.createdAt),
		description: proto.description,
		durationMinutes: proto.durationMinutes > 0 ? proto.durationMinutes : undefined,
		id: proto.id,
		recap: proto.recap,
		seriesId: proto.seriesId,
		startsAt: timestampDate(proto.startsAt),
		title: proto.title,
		updatedAt: timestampDate(proto.updatedAt),
	});
}

File: apps/api/internal/models/session.go
package model

import (
	"database/sql"
	"time"
)

type Session struct {
	ID              string
	CampaignID      string
	Title           string
	Description     sql.NullString
	SeriesID        sql.NullString
	ScheduledAt     time.Time
	CreatedAt       time.Time
	UpdatedAt       time.Time
	Recap           sql.NullString
	DurationMinutes int32
}

type CreateSessionRequest struct {
	CampaignID      string
	Title           string
	Description     sql.NullString
	SeriesID        sql.NullString
	ScheduledAt     time.Time
	DurationMinutes int32
}

type UpdateSessionRequest struct {
	ID          string
	Title       sql.NullString
	Description sql.NullString
	CampaignID  string
	Recap       sql.NullString
}
